src/encoding.rs

```diff
@@ -51,13 +52,32 @@ impl TryFrom<pkcs8::PrivateKeyInfoRef<'_>> for RsaPrivateKey {
             return Err(pkcs1::Error::Version.into());
         }
 
-        let n = BigUint::from_bytes_be(pkcs1_key.modulus.as_bytes());
-        let e = BigUint::from_bytes_be(pkcs1_key.public_exponent.as_bytes());
-        let d = BigUint::from_bytes_be(pkcs1_key.private_exponent.as_bytes());
-        let prime1 = BigUint::from_bytes_be(pkcs1_key.prime1.as_bytes());
-        let prime2 = BigUint::from_bytes_be(pkcs1_key.prime2.as_bytes());
+        let key_malformed = pkcs8::Error::KeyMalformed;
+
+        let bits =
+            u32::try_from(pkcs1_key.modulus.as_bytes().len()).map_err(|_| key_malformed)? * 8;
+
+        let n = BoxedUint::from_be_slice(pkcs1_key.modulus.as_bytes(), bits)
+            .map_err(|_| key_malformed)?;
+        let n = Option::from(Odd::new(n)).ok_or(key_malformed)?;
+
+        let bits_e = u32::try_from(pkcs1_key.public_exponent.as_bytes().len())
+            .map_err(|_| key_malformed)?
+            * 8;
+        let e = BoxedUint::from_be_slice(pkcs1_key.public_exponent.as_bytes(), bits_e)
+            .map_err(|_| key_malformed)?;
+        let e = Option::from(e).ok_or(key_malformed)?;
+
+        let d = BoxedUint::from_be_slice(pkcs1_key.private_exponent.as_bytes(), bits)
+            .map_err(|_| key_malformed)?;
+
+        let prime1 = BoxedUint::from_be_slice(pkcs1_key.prime1.as_bytes(), bits)
+            .map_err(|_| key_malformed)?;
+        let prime2 = BoxedUint::from_be_slice(pkcs1_key.prime2.as_bytes(), bits)
+            .map_err(|_| key_malformed)?;
```

Comment:
I'm not sure I quote understand why these conversions got more convoluted but perhaps there are upstream APIs which could be added to assist?

Reply:
Moving this to a method would clean it up a bunch. Call out to it and map_err from the call

Reply:
cleaned up a bit